common/dbconfig.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2021-present Pinterest. All Rights Reserved.
+//
+// @author Premkumar (prem@pinterest.com)
+//
+
+#include "common/dbconfig.h"
+#include "common/segment_utils.h"
+#include <utility>
+
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+
+DECLARE_int32(replicator_replication_mode);
+
+namespace common {
+
+DBConfigManager* DBConfigManager::get() {
+  static DBConfigManager instance;
+  return &instance;
+}
+
+DBConfigManager::DBConfigManager() : dbConfig_(new DBConfig()) {
+}
+
+bool DBConfigManager::loadJsonStr(const std::string& content) {
+  Json::Value root;
+  Json::Reader reader;
+  if (!reader.parse(content, root) || !root.isObject()) {
+    LOG(ERROR) << "Could not load config";
+    return false;
+  }
+
+  return loadJsonObject(root);
+}
+
+bool DBConfigManager::loadJsonObject(const Json::Value& root) {
+  try {
+    auto newConfig = parseConfig(root);
+    if (newConfig == nullptr) {
+      return false;
+    }
+
+    LOG(INFO) << "Successfully updated config";
+    std::atomic_exchange_explicit(&dbConfig_, newConfig, std::memory_order_release);
+    fDataLoaded_ = true;
+    return true;
+  } catch (std::exception& ex) {
+    LOG(ERROR) << "Exception from parseConfig: " << ex.what();
+    return false;
+  }
+}
+
+DBConfigPtr DBConfigManager::parseConfig(const Json::Value& root) {
```

Comment:
semantically, this should return a unique_ptr.

Reply:
how do we decide it's unique_ptr or a shared_ptr here, do we expect the parsed config here to be used in multiple places.

Also, since we need to atomically swap the `DBConfigPtr` upon config update, does that play a role here?

Reply:
A shared_ptr can be constructed from a unique_ptr. So we can still use shared_ptr for dbConfig_.

Semantically parseConfig() is producing a single copy of object, which is not being shared by anything that parseConfig() is aware of. So it's up to parseConfig()'s client to determine if they want to using the returned object via a unique_ptr (using it directly) or a shared_ptr (assign it to a shared_ptr dbConfig_).

Reply:
I've made this change in PR 565 ..